[JSC] JS Lexer should not do double-lookup for dispatch

Constellation opened the pull request:
#### a90e529da1cc5cc149294614ff6ca9fcd4568453
<pre>
[JSC] JS Lexer should not do double-lookup for dispatch
<a href="https://bugs.webkit.org/show_bug.cgi?id=296913">https://bugs.webkit.org/show_bug.cgi?id=296913</a>
<a href="https://rdar.apple.com/157533153">rdar://157533153</a>

Reviewed by Keith Miller.

For dispatching, we do not need to lookup the type from
typesOfLatin1Characters. It turned out that this load is really hot, and
we should just directly dispatch the switch based on the coming
character.
The important thing is,

1. We should list up all 0 - 255 cases explicitly to make sure that
   LChar is dispatched somewhere. This leads to efficient code
   generation from the compiler.
2. We should not add the characters outside of 0 - 255 to keep the
   dispatch super efficient. That&apos;s why we reuse white space characters
   for non-latin1 identifier dispatching.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer&lt;T&gt;::lexWithoutClearingLineTerminator):

Canonical link: <a href="https://commits.webkit.org/298259@main">https://commits.webkit.org/298259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95a6e7a2e1c69911f1aa94a9ca592f5496bfeb10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114837 "100 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25046 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120995 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65523 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c404d008-e931-4d59-9bf6-389f76e602b5) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35211 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43140 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87280 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8d50483a-b888-4273-ad17-1ec569fa2d36) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28063 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103118 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67672 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27244 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21239 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64659 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107180 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124194 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113391 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31266 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96087 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42217 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99307 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95872 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41064 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37888 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18389 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41722 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47236 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137605 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41279 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36791 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44595 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43023 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -1935,27 +1935,30 @@ JSTokenType Lexer<T>::lexWithoutClearingLineTerminator(JSToken* tokenRecord, Opt
-            type = CharacterLineTerminator;
-        else
-            type = CharacterInvalid;
+    if constexpr (sizeof(T) != sizeof(LChar)) {
```
Nit: `std::is_same_v`

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -1935,27 +1935,30 @@ JSTokenType Lexer<T>::lexWithoutClearingLineTerminator(JSToken* tokenRecord, Opt
 
     switch (type) {
-    case CharacterGreater:
+    case  62 /* > CharacterGreater */: {
```
Nit: Comment might be clearer as `/* 62 = '>' CharacterGreater */`. Ditto for the others below.

Constellation replied:
Changed.

kmiller68 reviewed (approved):
r=me

Constellation committed:
[JSC] JS Lexer should not do double-lookup for dispatch
https://bugs.webkit.org/show_bug.cgi?id=296913
rdar://157533153

Reviewed by Keith Miller.

For dispatching, we do not need to lookup the type from
typesOfLatin1Characters. It turned out that this load is really hot, and
we should just directly dispatch the switch based on the coming
character.
The important thing is,

1. We should list up all 0 - 255 cases explicitly to make sure that
   LChar is dispatched somewhere. This leads to efficient code
   generation from the compiler.
2. We should not add the characters outside of 0 - 255 to keep the
   dispatch super efficient. That's why we reuse white space characters
   for non-latin1 identifier dispatching.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer<T>::lexWithoutClearingLineTerminator):

Canonical link: https://commits.webkit.org/298259@main
